File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] leftPro = new int[nums.length];
        int[] rightPro = new int[nums.length];

        leftPro[0] = 1;

        for (int i = 1; i < nums.length; i++) {
            leftPro[i] = nums[i - 1] * leftPro[i - 1];
        }

        rightPro[nums.length - 1] = 1;

        for (int i = nums.length - 2; i >= 0; i--) {
            rightPro[i] = nums[i + 1] * rightPro[i + 1];
        }

        for (int i = 0; i < nums.length; i++) {
            nums[i] = leftPro[i] * rightPro[i];
        }

        return nums;
    }
}

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        zeroCount = 0
        product = 1

        for num in nums:
            if num != 0:
                product *= num
            else:
                zeroCount += 1

        if zeroCount > 1:
            return [0] * n
        elif zeroCount == 1:
            return [0 if num != 0 else product for num in nums]

        res = []

        for num in nums:
            res.append(product // num)

        return res

